encode.go

```diff
@@ -38,82 +39,134 @@ func (e Encoder) Encode(v interface{}) (err error) {
 	if err = e.encodeMapValueMaybe(); err != nil {
 		return
 	}
-	if v == nil {
+
+	// This type switch optimizes encoding of common value types, it prevents
+	// the use of reflection to identify the type of the value, which saves a
+	// dynamic memory allocation.
+	switch x := v.(type) {
+	case nil:
 		return e.Emitter.EmitNil()
-	}
-	return e.encode(reflect.ValueOf(v))
-}
 
-// EncodeBool uses e to encode the boolean value v.
-func (e Encoder) EncodeBool(v bool) (err error) {
```

Comment:
Indeed, these methods were here for optimization purposes but the new form is more generic. I know that we have no software that depended on those methods, there may be something out there that could be depended on it, in that case they'll need to adapt their code.

I'll add something to the readme explaining that change that needs to be made, but I don't think a deprecation period would be overkill for this case.